src/QsCompiler/DataStructures/DataTypes.fs

```diff
@@ -53,61 +53,118 @@ type QsNullable<'T> = // to avoid having to include the F# core in the C# part o
         | Some v -> Value v
         | None -> Null
 
+/// Operations for QsNullable<'T>.
+module QsNullable =
+    /// Applies 'f' to both nullable arguments if both have a Value. Returns Null otherwise.
+    let Map2 f a b =
+        match a, b with
+        | Value a, Value b -> f a b |> Value
+        | _ -> Null
 
 [<Struct>]
 type NonNullable<'T> = private Item of 'T with
     static member New value = if value = null then ArgumentNullException() |> raise else Item value
     member this.Value = this |> function | Item str -> str
 
 
-[<Struct>]
-type QsPositionInfo = {Line : int; Column: int} with 
-    static member Zero = {Line = 1; Column = 1}
-    static member CombinedRange (left : QsRangeInfo, right : QsRangeInfo) = 
-        match left, right with
-        | Value r1, Value r2 -> Value (fst r1, snd r2)
-        | _ -> Null
+/// A position in a text document.
+type Position = private Position of int * int with
+
+    /// The line number, where line zero is the first line in the document.
+    member this.Line = match this with Position (line, _) -> line
+
+    /// The column number, where column zero is the first column in the line.
+    member this.Column = match this with Position (_, column) -> column
+
+    /// Translates the first position by the amount of the second position. If the resulting position is on the same
+    /// line, the column numbers are added; otherwise, the second position's column number is used.
+    static member (+) (a : Position, b : Position) =
+        let line = a.Line + b.Line
+        let column = if b.Line = 0 then a.Column + b.Column else b.Column
+        Position (line, column)
+
+    /// Returns true if the positions have the same line and column numbers.
+    static member op_Equality (a : Position, b : Position) = a = b
+
+    /// Returns true if the second position occurs after the first position. 
+    static member op_GreaterThan (a : Position, b) = (a :> Position IComparable).CompareTo b > 0
+
+    /// <summary>
+    /// Creates a position.
+    /// </summary>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown if the line or column is negative.</exception>
+    static member Create line column =
```

Comment:
Maybe this is just my inexperience with F#, but why have a private constructor only to have a public static Create method? Couldn't you have just made the constructor public?

Reply:
It's not possible to define a custom constructor for an F# DU type.